compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp

```diff
@@ -427,4 +476,386 @@ LogicalResult splitLogicalObjectFifos(
+        l3ToL2DmaOps[i].getSourceMixedSizes();
+    unsigned j = 0, m = sourceOffsets.size();
+    // Traverse through the i-th L3->L2 Dma op's source offset/size to find a
+    // continuous sequence of 0 offset dims with size as 1.
+    while (j < m) {
+      std::optional<int64_t> constantOffset =
+          getConstantIntValue(sourceOffsets[j]);
+      if (!constantOffset || constantOffset.value() != 0) {
+        break;
+      }
+      std::optional<int64_t> constantSize = getConstantIntValue(sourceSizes[j]);
+      if (!constantSize || constantSize.value() != 1) {
+        break;
+      }
+      j++;
+    }
+    if (i == 0) {
+      maxSplitDimIndex = j;
+    } else if (maxSplitDimIndex != j) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "incompatible split dimensions across L3->L2\n");
+      return failure();
+    }
+  }
+  SmallVector<int64_t> splitDims(maxSplitDimIndex);
+  std::iota(splitDims.begin(), splitDims.end(), 0);
+  SmallVector<int64_t> nonSplitDims(maxSplitDimIndex);
+  std::iota(nonSplitDims.begin(), nonSplitDims.end(), splitDims.size());
+
+  // At this point it's nice to perhaps just sort the L3->L2 Dma ops based on
+  // the "overlapping" offsets. And we'll sort the corresponding L2->L1 Dma ops
+  // accordingly.
+  for (int64_t i = 1, n = l3ToL2DmaOps.size(); i < n; i++) {
+    DmaCpyNdOp currL3ToL2DmaOp = l3ToL2DmaOps[i];
+    DmaCpyNdOp currL2ToL1DmaOp = l2ToL1DmaOps[i];
+    int64_t j = i - 1;
+    while (j >= 0 && compareL3ToL2DmaPair(currL3ToL2DmaOp, l3ToL2DmaOps[j])) {
+      l3ToL2DmaOps[j + 1] = l3ToL2DmaOps[j];
+      l2ToL1DmaOps[j + 1] = l2ToL1DmaOps[j];
+      j--;
+    }
+    l3ToL2DmaOps[j + 1] = currL3ToL2DmaOp;
+    l2ToL1DmaOps[j + 1] = currL2ToL1DmaOp;
+  }
+
+  // Currently we have 4 cores so there are two pairs of DmaCpyNds to combine.
+  // TODO(avarma): Revisit this later when we want to target more no. of cores.
+  if (l3ToL2DmaOps.size() % 2 == 0) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "found uneven L3->L2 ops for combining\n");
+    return failure();
+  }
+
+  auto createL2ToL1ForReuse =
+      [](IRRewriter &rewriter, DmaCpyNdOp &l2ToL1DmaOp,
+         LogicalObjectFifoFromMemrefOp &reuseL1Buffer,
+         LogicalObjectFifoFromMemrefOp &reuseL2Buffer,
+         SmallVector<OpFoldResult> &newL2SourceOffsets) -> DmaCpyNdOp {
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPoint(l2ToL1DmaOp);
+    auto newL2ToL1DmaOp = rewriter.create<AMDAIE::DmaCpyNdOp>(
+        l2ToL1DmaOp.getLoc(), reuseL1Buffer,
+        l2ToL1DmaOp.getTargetMixedOffsets(), l2ToL1DmaOp.getTargetMixedSizes(),
+        l2ToL1DmaOp.getTargetMixedStrides(), reuseL2Buffer,
+        llvm::ArrayRef(newL2SourceOffsets), l2ToL1DmaOp.getSourceMixedSizes(),
+        l2ToL1DmaOp.getSourceMixedStrides());
+    rewriter.replaceOp(l2ToL1DmaOp, newL2ToL1DmaOp);
+    return newL2ToL1DmaOp;
+  };
+  for (unsigned i = 0, n = l3ToL2DmaOps.size(); i < n; i += 2) {
```

Comment:
Add a comment here to describe how you build the DMA chain + update the core ops